matrix_sdk/src/client.rs

```diff
@@ -1363,11 +1375,11 @@ impl Client {
     ) -> Result<register::Response> {
         info!("Registering to {}", self.homeserver().await);
 
-        #[cfg(not(feature = "appservice"))]
-        let config = None;
-
-        #[cfg(feature = "appservice")]
-        let config = Some(self.http_client.request_config.force_auth());
+        let config = if self.appservice_mode {
+            Some(self.http_client.request_config.force_auth())
+        } else {
+            None
+        };
```

Comment:
Given that the `appservice_mode` can only switched with `appservice` feature active this should be OK as runtime check